opensrp-opd/src/main/java/org/smartregister/opd/utils/VisitUtils.java

```diff
@@ -0,0 +1,241 @@
+
+        if (recordsUpdated < 1) {
+            abortTransaction();
+            throw new CheckInEventProcessException(String.format("Updating last interacted with for visit %s for base_entity_id %s in table %s failed"
+                    , visitId
+                    , event.getBaseEntityId()
+                    , tableName));
+        }
+
+        // Update FTS
+        CommonRepository commonrepository = OpdLibrary.getInstance().context().commonrepository(tableName);
+
+        ContentValues contentValues1 = new ContentValues();
+        contentValues1.put("last_interacted_with", lastInteractedWithDate);
+
+        boolean isUpdated = false;
+
+        if (commonrepository.isFts()) {
+            recordsUpdated = OpdLibrary.getInstance().getRepository().getWritableDatabase()
+                    .update(CommonFtsObject.searchTableName(tableName), contentValues, CommonFtsObject.idColumn + " = ?", new String[]{event.getBaseEntityId()});
+            isUpdated = recordsUpdated > 0;
+        }
+
+        if (!isUpdated) {
+            abortTransaction();
+            throw new CheckInEventProcessException(String.format("Updating last interacted with for visit %s for base_entity_id %s in table %s failed"
+                    , visitId
+                    , event.getBaseEntityId()
+                    , tableName));
+        }
+    }
+
+    public static void abortTransaction() {
+        if (OpdLibrary.getInstance().getRepository().getWritableDatabase().inTransaction()) {
+            OpdLibrary.getInstance().getRepository().getWritableDatabase().endTransaction();
+        }
+    }
+
+   /* public static String getDetailsValue(VisitDetail detail, String val) {
+        String clean_val = cleanString(val);
+        if (detail.getVisitKey().contains("date")) {
+            return getFormattedDate(getSourceDateFormat(), getSaveDateFormat(), clean_val);
+        }
+
+        return clean_val;
+    }
+*/
+    public static String getDetailsValue(VisitDetail detail, String val) {
+        String clean_val = cleanString(val);
+        if (detail.getVisitKey().contains("date") && StringUtils.isNotBlank(clean_val) && isValidDate(clean_val)) {
+            return getFormattedDate(getSourceDateFormat(), getSaveDateFormat(), clean_val);
+        }
+
+        return clean_val;
+    }
+
+    public static boolean isValidDate(String inDate) {
+        SimpleDateFormat dateFormat = new SimpleDateFormat("dd-MM-yyyy HH:mm:ss:ms");
+        dateFormat.setLenient(false);
+        try {
+            dateFormat.parse(inDate.trim());
+        } catch (ParseException pe) {
+            return false;
+        }
+        return true;
+    }
+
+        public static String getFormattedDate(SimpleDateFormat source_sdf, SimpleDateFormat dest_sdf, String value) {
+        try {
+            Date date = source_sdf.parse(value);
+            return dest_sdf.format(date);
+        } catch (Exception e) {
```

Comment:
```suggestion
        } catch (ParseException e) {
```

Or use a less generic exception.